### DISCUSSION TITLE
Generate cracks by duplicating nodes

### DISCUSSION THREAD

**Discussion by tfc1992** [idaholab/moose#16494] [Q&A Meshing] [answered]
Dear Moose user,
I am learning the phase field modeling of fracture using moose. And I want to create a pre-crack by duplicating the nodes. how do I realize it in moose?  any examples available?  Many thanks.

**Comment by ngrilli** [comment, marked as answer]
Dear @tfc1992
 
I believe this must be done in gmsh.
You can create "Point" and "Line" and "Line Loop" manually in the .geo file
You don't need to duplicate nodes, just define two different "Line" along the same segment
and define "Line Loop" by adding the two "Line" one after the other.
In this way the nodes on the two "Line" will be independent. 
I paste below and example of .geo file, but it's not for crack, you will need to modify it.
Hope this helps.

Best Regards,
Nicolò Grilli

//Points
Point(1) = {0, 0, 0, 10};
Point(2) = {100, 0, 0, 10};
Point(3) = {600, 0, 0, 10};
Point(4) = {700, 0, 0, 10};
Point(5) = {700, 500, 0, 10};
Point(6) = {500, 500, 0, 10};
Point(7) = {200, 500, 0, 10};
Point(8) = {0, 500, 0, 10};
Point(9) = {100, 200, 0, 10};
Point(10) = {200, 200, 0, 10};
Point(11) = {200, 300, 0, 10};
Point(12) = {500, 300, 0, 10};
Point(13) = {500, 200, 0, 10};
Point(14) = {600, 200, 0, 10};
Point(15) = {700, 200, 0, 10};
Point(16) = {0, 200, 0, 10};
Point(17) = {200, 0, 0, 10};
Point(18) = {500, 0, 0, 10};
Point(19) = {50, 0, 0, 10};
Point(20) = {50, 200, 0, 10};
Point(21) = {50, 500, 0, 10};

//Lines

Line(1) = {8, 21};
Line(2) = {21, 7};
Line(3) = {7, 6};
Line(4) = {6, 5};
Line(5) = {5, 15};
Line(6) = {15, 4};
Line(7) = {4, 3};
Line(8) = {3, 18};
Line(9) = {18, 17};
Line(10) = {17, 2};
Line(11) = {2, 19};
Line(12) = {19, 1};
Line(13) = {1, 16};
Line(14) = {16, 8};
Line(15) = {7, 11};
Line(16) = {11, 10};
Line(17) = {10, 9};
Line(18) = {9, 20};
Line(19) = {20, 16};
Line(20) = {6, 12};
Line(21) = {12, 13};
Line(22) = {13, 14};
Line(23) = {14, 15};
Line(24) = {3, 14};
Line(25) = {13, 18};
Line(26) = {2, 9};
Line(27) = {10, 17};
Line(28) = {11, 12};
Circle(29) = {9, 10, 11};
Circle(30) = {14, 13, 12};
Line(31) = {10, 13};
Line(32) = {19, 20};
Line(33) = {20, 21};

// Line Loops

Line Loop(101) = {1, -33, 19, 14};
Plane Surface(201) = {101};
Line Loop(102) = {2, 15, -29, 18, 33};
Plane Surface(202) = {102};
Line Loop(103) = {3, 20, -28, -15};
Plane Surface(203) = {103};
Line Loop(104) = {4, 5, -23, 30, -20};
Plane Surface(204) = {104};
Line Loop(105) = {16, 17, 29};
Plane Surface(205) = {105};
Line Loop(106) = {28, 21, -31, -16};
Plane Surface(206) = {106};
Line Loop(107) = {-30, -22, -21};
Plane Surface(207) = {107};
Line Loop(108) = {12, 13, -19, -32};
Plane Surface(208) = {108};
Line Loop(109) = {11, 32, -18, -26};
Plane Surface(209) = {109};
Line Loop(110) = {27, 10, 26, -17};
Plane Surface(210) = {110};
Line Loop(111) = {25, 9, -27, 31};
Plane Surface(211) = {111};
Line Loop(112) = {8, -25, 22, -24};
Plane Surface(212) = {112};
Line Loop(113) = {6, 7, 24, 23};
Plane Surface(213) = {113};

// Transfinite Line

Transfinite Line {7, 8, 10, 23, 22, 17, 16, 21, 29, 30} = 11 Using Progression 1;
Transfinite Line {6, 24, 25, 27, 26, 32, 13, 15, 20, 4} = 21 Using Progression 1;
Transfinite Line {9, 31, 28, 3, 5, 14, 33} = 31 Using Progression 1;
Transfinite Line {11, 12, 18, 19, 1} = 6 Using Progression 1;
Transfinite Line {2} = 16 Using Progression 1;

// Transfinite Surface

Transfinite Surface {201};
//Transfinite Surface {202};
Transfinite Surface {203};
//Transfinite Surface {204};
//Transfinite Surface {205};
Transfinite Surface {206};
//Transfinite Surface {207};
Transfinite Surface {208};
Transfinite Surface {209};
Transfinite Surface {210};
Transfinite Surface {211};
Transfinite Surface {212};
Transfinite Surface {213};

// Recombine Surface

Recombine Surface {201};
Recombine Surface {202};
Recombine Surface {203};
Recombine Surface {204};
Recombine Surface {205};
Recombine Surface {206};
Recombine Surface {207};
Recombine Surface {208};
Recombine Surface {209};
Recombine Surface {210};
Recombine Surface {211};
Recombine Surface {212};
Recombine Surface {213};

// Physical Line

Physical Line("top") = {1, 2, 3, 4};
Physical Line("right") = {5, 6};
Physical Line("bottom") = {7, 8, 9, 10, 11, 12};
Physical Line("left") = {13, 14};

// Physical Surface

Physical Surface("sylgard") = {201, 202, 203, 204, 208, 209, 213};
Physical Surface("hmx") = {205, 206, 207, 210, 211, 212};

**Reply by tfc1992** [reply]
Dear @ngrilli,  Very grateful for your help. I understand the method you proposed and I will try. Besides, I also found a plugin termed ''crack'' in Gmsh, but it didn't work. The following is a simple .geo file. Maybe I missed something.

//+
Point(1) = {0, 0, 0, 1.0};
//+
Point(2) = {1, 0, 0, 1.0};
//+
Point(3) = {1, 1, 0, 1.0};
//+
Point(4) = {-0, 1, 0, 1.0};
//+
Line(1) = {1, 2};
//+
Line(2) = {2, 3};
//+
Line(3) = {3, 4};
//+
Line(4) = {4, 1};
//+
Curve Loop(1) = {3, 4, 1, 2};
//+
Plane Surface(1) = {1};
//+
Physical Curve("top") = {3};
//+
Physical Curve("right") = {2};
//+
Physical Curve("left") = {4};
//+
Physical Curve("bottom") = {1};

//+
Transfinite Surface {1};
//+
Transfinite Surface {1} = {4, 3, 2, 1};
//+
Transfinite Curve {4, 2} = 5 Using Progression 1;
//+
Transfinite Curve {3, 1} = 5 Using Progression 1;
//+
Mesh 2;
RecombineMesh;
Plugin(Crack).Dimension=1;
Plugin(Crack).PhysicalGroup=5;
Plugin(Crack).OpenBoundaryPhysicalGroup=0;
Plugin(Crack).NormalX=0;
Plugin(Crack).NormalY=0;
Plugin(Crack).NormalZ=1;
Plugin(Crack).Run;

**Reply by ngrilli** [reply]
Dear @tfc1992 

I don't know about the crack plugin, I would just do that by hand.
For instance, in your file, you can create three more points:

Point(5) = {0, 0.5, 0, 1.0};
Point(6) = {0.5, 0.5, 0, 1.0};
Point(7) = {0, 0.5, 0, 1.0};

Then instead of connecting 4 to 1, you do:

Line(4) = {4, 5};
Line(5) = {5, 6};
Line(6) = {6, 7};

So line 5 and 6 will be different but located in the same positions.
When you generate the mesh, the nodes will be independent on those two lines
and if you apply vertical load, the crack will open.
Point 6 is the crack tip

This should work, let me know

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by tfc1992** [reply]
Dear @ngrilli  Thanks for your response.  I Ttried this method, but it can't achieve the result I want, like the Fig.6 in the following paper https://onlinelibrary.wiley.com/doi/full/10.1002/nme.6172,

**Reply by ngrilli** [reply]
Dear @tfc1992 
You can definitely reproduce that result in figure 6 with MOOSE.
It seems the crack surfaces in that figure are not coincident, so you can built a domain where the crack has already an opening
and you model the initial crack tip as a semi-sphere, you will see stress concentration at the tip and then the phase field =1 region will appear.
Please let us know which results you see in your simulations and share input file, otherwise it is difficult to help
Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by tfc1992** [reply]
Dear @ngrilli  Thank you for your help. Inspired by the method you proposed, I just got it in gmsh. Here is the .geo file. However, I haven't developed the moose input file yet. Actually, I am a new moose user. 

Point(1) = {0, 0, 0, 1.0};
//+
Point(2) = {0.5, 0, 0, 1.0};
//+
Point(3) = {1, 0, 0, 1.0};
//+
Point(4) = {1, 0.5, 0, 1.0};
//+
Point(5) = {1, 1, 0, 1.0};
//+
Point(6) = {0.5, 1, 0, 1.0};
//+
Point(7) = {0, 1, 0, 1.0};
//+
Point(8) = {0, 0.5, 0, 1.0};
//+
Point(9) = {0.5, 0.5, 0, 1.0};
//+
Point(10) = {-0, 0.5, 0, 1.0};
//+
Line(1)={1,2};
//+
Line(2) = {2, 3};
//+
Line(3) = {3, 4};
//+
Line(4) = {4, 5};
//+
Line(5) = {5, 6};
//+
Line(6) = {6, 7};
//+
Line(7) = {7, 8};
//+
Line(8) = {8, 9};
//+
Line(9) = {9, 10};
//+
Line(10) = {10, 1};
//+
Physical Curve("top") = {6, 5};
//+
Physical Curve("left") = {7, 10};
//+
Physical Curve("right") = {4, 3};
//+
Physical Curve("bottom") = {1, 2};
//+
Line(11) = {9, 6};
//+
Line(12) = {9, 4};
//+
Line(13) = {9, 2};
//+
Curve Loop(1) = {7, 8, 11, 6};
//+
Plane Surface(1) = {1};
//+
Curve Loop(2) = {5, -11, 12, 4};
//+
Plane Surface(2) = {2};
//+
Curve Loop(3) = {13, 2, 3, -12};
//+
Plane Surface(3) = {3};
//+
Curve Loop(4) = {9, 10, 1, -13};
//+
Plane Surface(4) = {4};
//+
Transfinite Curve {7, 11, 6, 8} = 100 Using Progression 1;
//+
Transfinite Curve {5, 11, 4, 12} =100 Using Progression 1;
//+
Transfinite Curve {3, 13, 12, 2} = 100 Using Progression 1;
//+
Transfinite Curve {9, 10, 13, 1} = 100 Using Progression 1;
//+
Transfinite Surface {1};
//+
Transfinite Surface {2};
//+
Transfinite Surface {3};
//+
Transfinite Surface {4};
//+
Recombine Surface {1};
//+
Recombine Surface {2};
//+
Recombine Surface {3};
//+
Recombine Surface {4};

**Reply by ngrilli** [reply]
Dear @tfc1992 

There are lot of ready input file for phase field fracture here:
https://github.com/idaholab/moose/tree/next/modules/combined/test/tests/phase_field_fracture

You just need to import your mesh using the FileMeshGenerator object
and assign "physical surfaces" in gmsh to apply boundary conditions in MOOSE

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by tfc1992** [reply]
Dear @ngrilli  Many thanks. I found it.

Best Regards,
Fucheng Tian

**Reply by RWTHLHK** [reply]
@tfc1992 Dear Fucheng, I also need the same mesh urgently. But it always reports an error when I import the mesh generated by your .geo file. It always says the boundary top can't be found... And I checked the mesh using peacock, the top, bottom, and so on are blocks not boundaries...

**Comment by dschwen** [comment]
An easy hack is to precrack along a boundary by constraining displacements only for part of the boundary (the remaining unconstrained part is the crack).

**Reply by tfc1992** [reply]
Dear @dschwen , Thanks for your response.  I tried this, but it won't work if the crack is not located at the boundary.

**Comment by jiangwen84** [comment]
To prescribe an initial crack for phase-field fracture in MOOSE, you can 
1. Use meshing software, such as gmsh and Cubit, to explicitly mesh the crack by either duplicating nodes or creating non-overlapping crack surfaces. 
2. You can use damage variable itself to prescribe the crack, and use either history variable approach or variational inequality solver to enforce the irreversibility of your initial crack.
3. Use XFEM capability to cut your mesh and represent the crack. Add the following blocks into your input. XFEM doesn't work with mesh adaptivity though. 

[XFEM]
  qrule = volfrac
  output_cut_plane = true
[]

[UserObjects]
  [./line_seg_cut_uo]
    type = LineSegmentCutUserObject
    cut_data = '0.5 1.0 0.5 0.5'   #cut_data = 'x0 y0 x1 y1' and  segment defined by point(x0,y0) and  point(x1 y1) #
    time_start_cut = 0.0
    time_end_cut = 0.0
  [../]
[]

**Reply by tfc1992** [reply]
Dear @jiangwen84  Thanks for your good suggestion. Under the help of  @ngrilli, I solved this problem in Gmsh. But, I will try the second method you provided.

Best Regards,
Fucheng Tian

**Comment by GregVernon** [comment]
@tfc1992,

@jiangwen84 provides several great methods.  If you are interested in his first suggestion, please know that you can get a free copy of Coreform Cubit Learn for non-commercial use, or a free-trial of Coreform Cubit for commercial use.  I recently helped a group from SDSU build a mesh for a geological fault - which is essentially a crack - by duplicating nodes as you've described.  I'd be happy to assist you with this approach if you're interested.

Link: [https://coreform.com/products/coreform-cubit/](https://coreform.com/products/coreform-cubit/ )

Best regards,

Greg Vernon
Director of Product Management
Coreform LLC
